File: backend/src/main/java/com/crm/service/ReportService.java
package com.crm.service;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.crm.dto.ReportFilterDTO;
import com.crm.dto.ReportQueryDTO;
import com.crm.dto.ReportResultDTO;
import com.crm.dto.ReportZeileDTO;
import com.crm.entity.enums.ChancePhase;
import com.crm.entity.enums.ReportDimension;
import com.crm.entity.enums.ReportMetrik;

import jakarta.persistence.EntityManager;
import jakarta.persistence.Query;

@Service
@Transactional(readOnly = true)
public class ReportService {

    private final EntityManager entityManager;

    public ReportService(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public ReportResultDTO executeReport(ReportQueryDTO query) {
        List<ReportMetrik> metriken = query.metriken();
        ReportDimension dimension = query.dimension();
        ReportFilterDTO filter = query.filter();

        boolean hasGewinnrate = metriken.contains(ReportMetrik.GEWINNRATE);
        List<ReportMetrik> aggregateMetriken = metriken.stream()
                .filter(m -> m != ReportMetrik.GEWINNRATE)
                .toList();

        String jpql = buildQuery(dimension, aggregateMetriken, filter);
        Query jpaQuery = entityManager.createQuery(jpql);
        setParameters(jpaQuery, filter);

        @SuppressWarnings("unchecked")
        List<Object[]> rows = jpaQuery.getResultList();

        List<ReportZeileDTO> zeilen = new ArrayList<>();
        for (Object[] row : rows) {
            String label = buildLabel(row, dimension);
            Long id = buildId(row, dimension);

            int offset = dimensionColumnCount(dimension);
            Map<String, Number> werte = new LinkedHashMap<>();

            int metrikIdx = 0;
            for (ReportMetrik metrik : aggregateMetriken) {
                Number val = row[offset + metrikIdx] != null
                        ? BigDecimal.valueOf(((Number) row[offset + metrikIdx]).doubleValue())
                        : null;
                werte.put(metrik.name(), val);
                metrikIdx++;
            }

            zeilen.add(new ReportZeileDTO(label, id, werte));
        }

        if (hasGewinnrate) {
            calculateGewinnrate(zeilen, dimension, filter);
        }

        return new ReportResultDTO(dimension, metriken, zeilen);
    }

    private String buildQuery(ReportDimension dimension, List<ReportMetrik> metriken,
            ReportFilterDTO filter) {
        StringBuilder sb = new StringBuilder("SELECT ");
        sb.append(dimensionSelect(dimension));

        for (ReportMetrik metrik : metriken) {
            sb.append(", ").append(metrikSelect(metrik));
        }

        sb.append(" FROM Chance c");
        sb.append(" LEFT JOIN c.firma f");
        sb.append(" LEFT JOIN c.kontaktPerson p");
        sb.append(whereClause(filter, dimension));
        sb.append(" GROUP BY ").append(dimensionGroupBy(dimension));
        sb.append(" ORDER BY ").append(dimensionOrderBy(dimension));

        return sb.toString();
    }

    private String dimensionSelect(ReportDimension dimension) {
        return switch (dimension) {
            case PHASE -> "c.phase";
            case FIRMA -> "f.id, f.name";
            case PERSON -> "p.id, p.firstName, p.lastName";
            case MONAT -> "YEAR(c.erwartetesDatum), MONTH(c.erwartetesDatum)";
            case QUARTAL -> "YEAR(c.erwartetesDatum), QUARTER(c.erwartetesDatum)";
            case JAHR -> "YEAR(c.erwartetesDatum)";
        };
    }

    private String dimensionGroupBy(ReportDimension dimension) {
        return switch (dimension) {
            case PHASE -> "c.phase";
            case FIRMA -> "f.id, f.name";
            case PERSON -> "p.id, p.firstName, p.lastName";
            case MONAT -> "YEAR(c.erwartetesDatum), MONTH(c.erwartetesDatum)";
            case QUARTAL -> "YEAR(c.erwartetesDatum), QUARTER(c.erwartetesDatum)";
            case JAHR -> "YEAR(c.erwartetesDatum)";
        };
    }

    private String dimensionOrderBy(ReportDimension dimension) {
        return switch (dimension) {
            case PHASE -> "c.phase";
            case FIRMA -> "f.name";
            case PERSON -> "p.lastName, p.firstName";
            case MONAT -> "YEAR(c.erwartetesDatum), MONTH(c.erwartetesDatum)";
            case QUARTAL -> "YEAR(c.erwartetesDatum), QUARTER(c.erwartetesDatum)";
            case JAHR -> "YEAR(c.erwartetesDatum)";
        };
    }

    private int dimensionColumnCount(ReportDimension dimension) {
        return switch (dimension) {
            case PHASE, JAHR -> 1;
            case FIRMA, MONAT, QUARTAL -> 2;
            case PERSON -> 3;
        };
    }

    private String metrikSelect(ReportMetrik metrik) {
        return switch (metrik) {
            case ANZAHL -> "COUNT(c)";
            case SUMME_WERT -> "SUM(c.wert)";
            case DURCHSCHNITT_WERT -> "AVG(c.wert)";
            case GEWICHTETER_WERT -> "SUM(c.wert * c.wahrscheinlichkeit / 100.0)";
            case GEWINNRATE -> throw new IllegalArgumentException("GEWINNRATE is handled separately");
        };
    }

    private String whereClause(ReportFilterDTO filter, ReportDimension dimension) {
        List<String> conditions = new ArrayList<>();

        // Exclude NULLs for nullable dimension columns
        if (dimension == ReportDimension.PERSON) {
            conditions.add("c.kontaktPerson IS NOT NULL");
        } else if (dimension == ReportDimension.MONAT || dimension == ReportDimension.QUARTAL
                || dimension == ReportDimension.JAHR) {
            conditions.add("c.erwartetesDatum IS NOT NULL");
        }

        if (filter != null) {
            if (filter.phasen() != null && !filter.phasen().isEmpty()) {
                conditions.add("c.phase IN :phasen");
            }
            if (filter.datumVon() != null) {
                conditions.add("c.erwartetesDatum >= :datumVon");
            }
            if (filter.datumBis() != null) {
                conditions.add("c.erwartetesDatum <= :datumBis");
            }
        }

        if (conditions.isEmpty()) {
            return "";
        }

        return " WHERE " + String.join(" AND ", conditions);
    }

    private void setParameters(Query jpaQuery, ReportFilterDTO filter) {
        if (filter == null) {
            return;
        }

        if (filter.phasen() != null && !filter.phasen().isEmpty()) {
            jpaQuery.setParameter("phasen", filter.phasen());
        }
        if (filter.datumVon() != null) {
            jpaQuery.setParameter("datumVon", filter.datumVon());
        }
        if (filter.datumBis() != null) {
            jpaQuery.setParameter("datumBis", filter.datumBis());
        }
    }

    private String buildLabel(Object[] row, ReportDimension dimension) {
        return switch (dimension) {
            case PHASE -> ((ChancePhase) row[0]).name();
            case FIRMA -> (String) row[1];
            case PERSON -> row[1] + " " + row[2];
            case MONAT -> String.format("%d-%02d", ((Number) row[0]).intValue(), ((Number) row[1]).intValue());
            case QUARTAL -> String.format("Q%d %d", ((Number) row[1]).intValue(), ((Number) row[0]).intValue());
            case JAHR -> String.valueOf(((Number) row[0]).intValue());
        };
    }

    private Long buildId(Object[] row, ReportDimension dimension) {
        return switch (dimension) {
            case FIRMA -> row[0] != null ? ((Number) row[0]).longValue() : null;
            case PERSON -> row[0] != null ? ((Number) row[0]).longValue() : null;
            default -> null;
        };
    }

    private void calculateGewinnrate(List<ReportZeileDTO> zeilen, ReportDimension dimension,
            ReportFilterDTO filter) {
        // Check if filter excludes GEWONNEN or VERLOREN
        if (filter != null && filter.phasen() != null && !filter.phasen().isEmpty()) {
            boolean hasGewonnen = filter.phasen().contains(ChancePhase.GEWONNEN);
            boolean hasVerloren = filter.phasen().contains(ChancePhase.VERLOREN);
            if (!hasGewonnen || !hasVerloren) {
                // Cannot calculate Gewinnrate without both phases
                for (ReportZeileDTO zeile : zeilen) {
                    zeile.werte().put(ReportMetrik.GEWINNRATE.name(), null);
                }
                return;
            }
        }

        // For PHASE dimension, calculate inline
        if (dimension == ReportDimension.PHASE) {
            for (ReportZeileDTO zeile : zeilen) {
                if ("GEWONNEN".equals(zeile.label()) || "VERLOREN".equals(zeile.label())) {
                    // Will be calculated after iterating
                }
                zeile.werte().put(ReportMetrik.GEWINNRATE.name(), null);
            }

            // Find Gewonnen and Verloren counts
            long gewonnen = 0;
            long verloren = 0;
            for (ReportZeileDTO zeile : zeilen) {
                Number anzahl = zeile.werte().get(ReportMetrik.ANZAHL.name());
                if (anzahl == null) {
                    // Need a separate count query
                    break;
                }
                if ("GEWONNEN".equals(zeile.label())) {
                    gewonnen = anzahl.longValue();
                } else if ("VERLOREN".equals(zeile.label())) {
                    verloren = anzahl.longValue();
                }
            }

            BigDecimal rate = (gewonnen + verloren) > 0
                    ? BigDecimal.valueOf((double) gewonnen / (gewonnen + verloren) * 100.0)
                    : null;

            // Set Gewinnrate for all phase rows (same global rate)
            for (ReportZeileDTO zeile : zeilen) {
                zeile.werte().put(ReportMetrik.GEWINNRATE.name(), rate);
            }
            return;
        }

        // For other dimensions: run a separate query per group
        String gewinnrateJpql = buildGewinnrateQuery(dimension, filter);
        Query query = entityManager.createQuery(gewinnrateJpql);
        setParameters(query, filter);

        @SuppressWarnings("unchecked")
        List<Object[]> results = query.getResultList();

        // Build a lookup: label -> gewinnrate
        Map<String, BigDecimal> rateByLabel = new LinkedHashMap<>();
        for (Object[] row : results) {
            String label = buildLabel(row, dimension);
            int offset = dimensionColumnCount(dimension);
            long gewonnen = ((Number) row[offset]).longValue();
            long verloren = ((Number) row[offset + 1]).longValue();
            BigDecimal rate = (gewonnen + verloren) > 0
                    ? BigDecimal.valueOf((double) gewonnen / (gewonnen + verloren) * 100.0)
                    : null;
            rateByLabel.put(label, rate);
        }

        for (ReportZeileDTO zeile : zeilen) {
            zeile.werte().put(ReportMetrik.GEWINNRATE.name(), rateByLabel.get(zeile.label()));
        }
    }

    private String buildGewinnrateQuery(ReportDimension dimension, ReportFilterDTO filter) {
        StringBuilder sb = new StringBuilder("SELECT ");
        sb.append(dimensionSelect(dimension));
        sb.append(", SUM(CASE WHEN c.phase = com.crm.entity.enums.ChancePhase.GEWONNEN THEN 1 ELSE 0 END)");
        sb.append(", SUM(CASE WHEN c.phase = com.crm.entity.enums.ChancePhase.VERLOREN THEN 1 ELSE 0 END)");
        sb.append(" FROM Chance c");
        sb.append(" LEFT JOIN c.firma f");
        sb.append(" LEFT JOIN c.kontaktPerson p");

        List<String> conditions = new ArrayList<>();
        conditions.add("(c.phase = com.crm.entity.enums.ChancePhase.GEWONNEN OR c.phase = com.crm.entity.enums.ChancePhase.VERLOREN)");
        if (dimension == ReportDimension.PERSON) {
            conditions.add("c.kontaktPerson IS NOT NULL");
        } else if (dimension == ReportDimension.MONAT || dimension == ReportDimension.QUARTAL
                || dimension == ReportDimension.JAHR) {
            conditions.add("c.erwartetesDatum IS NOT NULL");
        }
        if (filter != null) {
            if (filter.datumVon() != null) {
                conditions.add("c.erwartetesDatum >= :datumVon");
            }
            if (filter.datumBis() != null) {
                conditions.add("c.erwartetesDatum <= :datumBis");
            }
        }

        sb.append(" WHERE ").append(String.join(" AND ", conditions));
        sb.append(" GROUP BY ").append(dimensionGroupBy(dimension));
        sb.append(" ORDER BY ").append(dimensionOrderBy(dimension));

        return sb.toString();
    }
}

File: backend/src/services/reportService.ts
import { sqlite } from '../config/db.js';
import { ValidationError } from '../utils/errors.js';
import { CHANCE_PHASE, REPORT_DIMENSION, REPORT_METRIK } from '../db/schema/enums.js';
import type { ReportDimension, ReportMetrik } from '../db/schema/enums.js';
import type { ReportQueryDTO } from '../utils/validation.js';

export interface ReportZeile {
  label: string;
  id: string | number | null;
  werte: Record<string, number>;
}

export interface ReportResult {
  dimension: string;
  metriken: string[];
  zeilen: ReportZeile[];
}

// SQLite date function translations
function yearExpr(col: string): string {
  return `CAST(strftime('%Y', ${col}) AS INTEGER)`;
}

function monthExpr(col: string): string {
  return `CAST(strftime('%m', ${col}) AS INTEGER)`;
}

function quarterExpr(col: string): string {
  return `((CAST(strftime('%m', ${col}) AS INTEGER) - 1) / 3 + 1)`;
}

// Build dimension SELECT + GROUP BY + label logic
function buildDimensionParts(dimension: ReportDimension): {
  selectDim: string;
  groupBy: string;
  idExpr: string;
  labelExpr: string;
  join: string;
} {
  switch (dimension) {
    case 'PHASE':
      return {
        selectDim: 'c.phase AS dim_phase',
        groupBy: 'c.phase',
        idExpr: 'c.phase',
        labelExpr: 'c.phase',
        join: '',
      };
    case 'FIRMA':
      return {
        selectDim: 'c.firmaId AS dim_firmaId, f.name AS dim_firmaName',
        groupBy: 'c.firmaId',
        idExpr: 'c.firmaId',
        labelExpr: 'COALESCE(f.name, CAST(c.firmaId AS TEXT))',
        join: 'LEFT JOIN firma f ON c.firmaId = f.id',
      };
    case 'PERSON':
      return {
        selectDim:
          'c.kontaktPersonId AS dim_personId, p.firstName AS dim_firstName, p.lastName AS dim_lastName',
        groupBy: 'c.kontaktPersonId',
        idExpr: 'c.kontaktPersonId',
        labelExpr:
          "COALESCE(p.firstName || ' ' || p.lastName, CAST(c.kontaktPersonId AS TEXT), 'Unbekannt')",
        join: 'LEFT JOIN person p ON c.kontaktPersonId = p.id',
      };
    case 'MONAT':
      return {
        selectDim: `${yearExpr('c.createdAt')} AS dim_year, ${monthExpr('c.createdAt')} AS dim_month`,
        groupBy: `${yearExpr('c.createdAt')}, ${monthExpr('c.createdAt')}`,
        idExpr: `${yearExpr('c.createdAt')} || '-' || printf('%02d', ${monthExpr('c.createdAt')})`,
        labelExpr: `${yearExpr('c.createdAt')} || '-' || printf('%02d', ${monthExpr('c.createdAt')})`,
        join: '',
      };
    case 'QUARTAL':
      return {
        selectDim: `${yearExpr('c.createdAt')} AS dim_year, ${quarterExpr('c.createdAt')} AS dim_quarter`,
        groupBy: `${yearExpr('c.createdAt')}, ${quarterExpr('c.createdAt')}`,
        idExpr: `'Q' || ${quarterExpr('c.createdAt')} || ' ' || ${yearExpr('c.createdAt')}`,
        labelExpr: `'Q' || ${quarterExpr('c.createdAt')} || ' ' || ${yearExpr('c.createdAt')}`,
        join: '',
      };
    case 'JAHR':
      return {
        selectDim: `${yearExpr('c.createdAt')} AS dim_year`,
        groupBy: yearExpr('c.createdAt'),
        idExpr: yearExpr('c.createdAt'),
        labelExpr: yearExpr('c.createdAt'),
        join: '',
      };
    default:
      throw new ValidationError(`Unbekannte Dimension: ${dimension as string}`);
  }
}

// Build metric SELECT expressions
function buildMetrikSelect(metriken: ReportMetrik[]): string[] {
  return metriken
    .filter((m) => m !== 'GEWINNRATE') // handled via post-processing
    .map((metrik) => {
      switch (metrik) {
        case 'ANZAHL':
          return 'COUNT(*) AS metrik_ANZAHL';
        case 'GESAMTWERT':
          return 'COALESCE(SUM(c.wert), 0) AS metrik_GESAMTWERT';
        case 'DURCHSCHNITTSWERT':
          return 'COALESCE(AVG(c.wert), 0) AS metrik_DURCHSCHNITTSWERT';
        case 'GEWICHTETER_WERT':
          return 'COALESCE(SUM(c.wert * c.wahrscheinlichkeit / 100.0), 0) AS metrik_GEWICHTETER_WERT';
        default:
          return '';
      }
    })
    .filter(Boolean);
}

export const reportService = {
  executeReport(query: ReportQueryDTO): ReportResult {
    const dimension = query.dimension as ReportDimension;
    const metriken = query.metriken as ReportMetrik[];

    if (!(REPORT_DIMENSION as readonly string[]).includes(dimension)) {
      throw new ValidationError(`Ungültige Dimension: ${dimension}`);
    }
    for (const m of metriken) {
      if (!(REPORT_METRIK as readonly string[]).includes(m)) {
        throw new ValidationError(`Ungültige Metrik: ${m}`);
      }
    }

    const dimParts = buildDimensionParts(dimension);
    const metrikSelects = buildMetrikSelect(metriken);

    // Build WHERE conditions
    const whereClauses: string[] = [];
    const params: unknown[] = [];

    const filter = query.filter;
    if (filter?.phasen && filter.phasen.length > 0) {
      const inPlaceholders = filter.phasen.map((_, i) => `?`).join(', ');
      whereClauses.push(`c.phase IN (${inPlaceholders})`);
      params.push(...filter.phasen);
    }
    if (filter?.datumVon) {
      whereClauses.push('c.createdAt >= ?');
      params.push(filter.datumVon);
    }
    if (filter?.datumBis) {
      whereClauses.push('c.createdAt <= ?');
      params.push(filter.datumBis);
    }

    const where = whereClauses.length > 0 ? `WHERE ${whereClauses.join(' AND ')}` : '';

    const allSelects = [dimParts.selectDim, dimParts.labelExpr + ' AS dim_label', dimParts.idExpr + ' AS dim_id', ...metrikSelects];

    const sql = `
      SELECT ${allSelects.join(', ')}
      FROM chance c
      ${dimParts.join}
      ${where}
      GROUP BY ${dimParts.groupBy}
      ORDER BY ${dimParts.groupBy}
    `;

    const rows = sqlite.prepare(sql).all(...params) as Record<string, unknown>[];

    // Post-process GEWINNRATE if requested
    const needsGewinnrate = metriken.includes('GEWINNRATE');
    let gewinnrateMap: Map<string, number> = new Map();

    if (needsGewinnrate) {
      // Build separate query for win rate
      const gwSql = `
        SELECT ${dimParts.idExpr} AS dim_id,
               SUM(CASE WHEN c.phase = 'GEWONNEN' THEN 1 ELSE 0 END) AS gewonnen,
               SUM(CASE WHEN c.phase IN ('GEWONNEN','VERLOREN') THEN 1 ELSE 0 END) AS abgeschlossen
        FROM chance c
        ${dimParts.join}
        ${where}
        GROUP BY ${dimParts.groupBy}
      `;
      const gwRows = sqlite.prepare(gwSql).all(...params) as {
        dim_id: unknown;
        gewonnen: number;
        abgeschlossen: number;
      }[];
      for (const r of gwRows) {
        const key = String(r.dim_id ?? '');
        const rate =
          Number(r.abgeschlossen) > 0
            ? (Number(r.gewonnen) / Number(r.abgeschlossen)) * 100
            : 0;
        gewinnrateMap.set(key, Math.round(rate * 100) / 100);
      }
    }

    const zeilen: ReportZeile[] = rows.map((row) => {
      const label = String(row['dim_label'] ?? '');
      const id = row['dim_id'] as string | number | null;
      const werte: Record<string, number> = {};

      for (const metrik of metriken) {
        if (metrik === 'GEWINNRATE') {
          werte[metrik] = gewinnrateMap.get(String(id ?? '')) ?? 0;
        } else {
          const key = `metrik_${metrik}`;
          werte[metrik] = Number(row[key] ?? 0);
        }
      }

      return { label, id, werte };
    });

    // For PHASE dimension, fill in missing phases with zeros
    if (dimension === 'PHASE') {
      const existingPhases = new Set(zeilen.map((z) => z.label));
      for (const phase of CHANCE_PHASE) {
        if (!existingPhases.has(phase)) {
          const werte: Record<string, number> = {};
          for (const m of metriken) {
            werte[m] = 0;
          }
          zeilen.push({ label: phase, id: phase, werte });
        }
      }
      // Sort in CHANCE_PHASE order
      zeilen.sort(
        (a, b) =>
          (CHANCE_PHASE as readonly string[]).indexOf(a.label) -
          (CHANCE_PHASE as readonly string[]).indexOf(b.label)
      );
    }

    return { dimension, metriken, zeilen };
  },
};
